cmd/kubeadm/app/util/config/common.go

```diff
@@ -91,7 +91,7 @@ func validateSupportedVersion(gv schema.GroupVersion, allowDeprecated, allowExpe
 	}
 
 	if _, present := deprecatedAPIVersions[gvString]; present && !allowDeprecated {
-		klog.Warningf("your configuration file uses a deprecated API spec: %q. Please use 'kubeadm config migrate --old-config old.yaml --new-config new.yaml', which will write the new, similar spec using a newer API version.", gv)
+		klog.Warningf("your configuration file uses a deprecated API spec: %q. Please use 'kubeadm config migrate --old-config old.yaml --new-config new.yaml', which will write the new, similar spec using a newer API version.", gv.String())
 	}
 
 	if _, present := experimentalAPIVersions[gvString]; present && !allowExperimental {
```

Comment:
_just picking some line in common.go to comment on_

i think we are missing the following:
- support for validation and migration of reset kinds "kubeadm config validation/migrate" see how join/init configuration are handled in common.go and in the respective "kubeadm config" cmd/config.go functions.
- printing of reset defaults. right now we have "kubeadm config print ..." and we can print init / join defaults. we should add reset defaults there as well. in the future upgrade defaults too.

Reply:
+1, How about break in up to 2/3 PRs and leave these in another review? I will create a TODO items on the https://github.com/kubernetes/kubeadm/issues/2890, so that we won't miss this.

Reply:
updated the https://github.com/kubernetes/kubeadm/issues/2890 to have a TODO item for this.

Reply:
sgtm